### PR TITLE
FEATURE: Mark all chat channels read with a shortcut

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit adds a keyboard shortcut (Shift+ESC) for chat which marks all
of the chat channels that the user is currently a following member of as read,
updating their `last_read_message_id`. This is done via a new service.

It also includes some refactors and controller changes:

* The old mark message read route from `ChatController` is now supplanted
  by the `Chat::Api::ReadsController#update` route.
* The new controller can handle either marking a single or all messages read,
  and uses the correct service based on the route and params.
* The `UpdateUserLastRead` service is now used (it wasn't before), and has been slightly
  updated to just use the guardian user ID.

### CODE REVIEW

File: plugins/chat/app/controllers/chat/api/tracking_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class Chat::Api::TrackingController < Chat::ApiController
```
Comment (jjaffeux) [inline plugins/chat/app/controllers/chat/api/tracking_controller.rb:3]: I would suggest this as route to update last read:

```
PUT /channels/:id/read/:message_id
```

So we would have `Chat::Api::ChannelReadsController#update`

File: plugins/chat/app/controllers/api/chat_tracking_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class Chat::Api::ChatTrackingController < Chat::Api
+  def read
+    params.permit(:channel_id, :message_id)
+
+    channel_id = params[:channel_id]
+    message_id = params[:message_id]
+
+    if channel_id.present? && message_id.present?
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/app/controllers/api/chat_tracking_controller.rb]: Expanding on this from my other comment: https://github.com/discourse/discourse/pull/20629/files#r1141082660

I would prefer two routes yes, I would suggest:

```
PUT /channels/read/
PUT /channels/:id/read/:message_id
```

My feeling is that we will soon move to something like this:

```
PUT /channels/read/:timestamp
PUT /channels/:id/read/:timestamp
```

We would probably still return `last_read_message_id` for quick check in frontend, but if we found one unread message we would request with a timestamp.

File: plugins/chat/config/routes.rb [hunk 3]
```diff
@@ -7,7 +7,7 @@
     get "/channels/me" => "current_user_channels#index"
     post "/channels" => "channels#create"
     delete "/channels/:channel_id" => "channels#destroy"
-    put "/channels/:channel_id" => "channels#update"
+    put "/channels/:channel_id" => "channels#update", :constraints => { channel_id: /\d+/ }
```
Comment (martin-brennan) [inline plugins/chat/config/routes.rb:10]: This is required because otherwise it conflicts with the `channels/read` route

**Comment by martin-brennan** [comment]
@jjaffeux I think this is good for another look now, changed the routes 👍 Though I haven't made any changes for the timestamp thing, I think we can do that separately?

File: plugins/chat/config/routes.rb (hunk 3 above)
Comment (jjaffeux) [reply plugins/chat/config/routes.rb:10]: I think you just have to put read route first no?

File: plugins/chat/config/routes.rb [hunk 4]
```diff
@@ -17,6 +17,8 @@
     post "/channels/:channel_id/memberships/me" => "channels_current_user_membership#create"
     put "/channels/:channel_id/notifications-settings/me" =>
           "channels_current_user_notifications_settings#update"
+    put "/channels/read/" => "reads#update_all"
```
Comment (jjaffeux) [inline plugins/chat/config/routes.rb:20]: Sorry Im nitpicking here, but the `_all` is not needed IMO

File: plugins/chat/assets/javascripts/discourse/services/chat-api.js [hunk 5]
```diff
@@ -282,6 +282,23 @@ export default class ChatApi extends Service {
...
+  /**
+   * Marks messages for a user's chat channel memberships as read. If no
+   * channel ID and no message ID are provided, then all of the user's
+   * followed chat channel memberships will be marked as read.
+   *
+   * @param {number} channelId - The ID of the channel for the message being marked as read.
+   * @param {number} messageId - The ID of the message being marked as read.
+   * @returns {Promise}
+   */
+  updateCurrentUserRead({ channelId, messageId } = {}) {
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/services/chat-api.js:294]: Maybe we go with two functions here too?

```
markChannelAsRead(messageId) {} // maybe messageId optional to just let backend find the most recent message
markAllChannelsAsRead() {}
```

File: plugins/chat/app/services/chat/update_user_last_read.rb [hunk 6]
```diff
@@ -4,76 +4,76 @@ module Chat
...
+    include ::Service::Base
 
-    # @!method call(user_id:, channel_id:, message_id:, guardian:)
-    #   @param [Integer] user_id
+    # @!method call(channel_id:, message_id:, guardian:)
     #   @param [Integer] channel_id
     #   @param [Integer] message_id
     #   @param [Guardian] guardian
-    #   @return [Service::Base::Context]
+    #   @return [Chat::Service::Base::Context]
```
Comment (jjaffeux) [inline plugins/chat/app/services/chat/update_user_last_read.rb:16]: I think this is it now  no?


```suggestion
    #   @return [Service::Base::Context]
```

File: plugins/chat/app/services/chat/update_user_last_read.rb [hunk 7]
```diff
@@ -4,76 +4,76 @@ module Chat
...
-        .where(user: membership.user)
+        .where(user: active_membership.user)
         .where(read: false)
         .joins("INNER JOIN chat_mentions ON chat_mentions.notification_id = notifications.id")
         .joins("INNER JOIN chat_messages ON chat_mentions.chat_message_id = chat_messages.id")
-        .where("chat_messages.id <= ?", message_id)
-        .where("chat_messages.chat_channel_id = ?", membership.chat_channel.id)
+        .where("chat_messages.id <= ?", contract.message_id)
+        .where("chat_messages.chat_channel_id = ?", active_membership.chat_channel.id)
         .update_all(read: true)
```
Comment (jjaffeux) [inline plugins/chat/app/services/chat/update_user_last_read.rb:72]: Should we extract this as an action (it's very similar to the one in mark_all_user_channels...)? Fine to be a in a different PR

File: plugins/chat/config/routes.rb (hunk 4 above)
Comment (martin-brennan) [reply plugins/chat/config/routes.rb:20]: But they both go to the same controller, are you saying they should still both be in the same endpoint with the switching based on params like I had before?

File: plugins/chat/assets/javascripts/discourse/services/chat-api.js (hunk 5 above)
Comment (martin-brennan) [reply plugins/chat/assets/javascripts/discourse/services/chat-api.js:294]: Fair enough, will change

File: plugins/chat/app/services/chat/update_user_last_read.rb (hunk 6 above)
Comment (martin-brennan) [reply plugins/chat/app/services/chat/update_user_last_read.rb:16]: I guess, there are a few places that still refer to that namespace (I guess from the zeitwerk move?) so I will update them all

File: plugins/chat/app/services/chat/update_user_last_read.rb (hunk 7 above)
Comment (martin-brennan) [reply plugins/chat/app/services/chat/update_user_last_read.rb:72]: True, will extract in this same PR :+1:

File: plugins/chat/config/routes.rb (hunk 3 above)
Comment (martin-brennan) [reply plugins/chat/config/routes.rb:10]: Yes that seems to work, thanks :+1:

**Comment by martin-brennan** [comment]
@jjaffeux fixed the suggestions but just have one more question if you could please take a look

File: plugins/chat/config/routes.rb (hunk 4 above)
Comment (jjaffeux) [reply plugins/chat/config/routes.rb]: I was thinking two controllers... but you are right that might be weird, ok :+1:

Comment (martin-brennan) [reply plugins/chat/config/routes.rb]: Sweet, will leave as-is

File: plugins/chat/app/controllers/chat/api/reads_controller.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class Chat::Api::ReadsController < Chat::ApiController
+  def update
+    params.require(:channel_id)
```
Comment (oblakeerickson) [inline plugins/chat/app/controllers/chat/api/reads_controller.rb:5]: no biggy, but isn't the single line params.require([:channel_id, :message_id]) equivalent?

**Review by oblakeerickson — APPROVED** [review]

File: plugins/chat/app/controllers/chat/api/reads_controller.rb (hunk 8 above)
Comment (martin-brennan) [reply plugins/chat/app/controllers/chat/api/reads_controller.rb:5]: Ah I didn't even realise you could give it an array :laughing: I always tried `params.require(:x, :y)` and got annoyed when it wouldn't work. Will fix
